scalameta/tokenizers/shared/src/main/scala/scala/meta/internal/tokenizers/LegacyScanner.scala

```diff
@@ -232,9 +232,22 @@ class LegacyScanner(input: Input, dialect: Dialect) {
       }
     }
 
+    /**
+     * The algorithm for calculating unicode is as follows:
+     *   - `ch | 0x10000` gets us a new binary number: 1 at 17 position, then zeroes and then the
+     *     actual char number.
+     *   - `Integer.toHexString(ch | 0x10000)` gets us the hexagonal unicode plus the one from 17
+     *     position
+     *   - `Integer.toHexString(ch | 0x10000).substring(1)` drops that added one
+     *
+     * This way we don't have to prefix with 0s since they will be included in the hex number
+     * string.
+     */
```

Comment:
ah, now i understand it! well, in that case, this assumes that `ch` fits into 16 bits, and it doesn't: https://stackoverflow.com/questions/27415935/does-unicode-have-a-defined-maximum-number-of-code-points

especially since illegal characters likely are larger.

here are two alternatives i can propose (need to write a short test, of course):
```
/**
 * Prints exactly the number of bytes in `ch` and has no extra memory allocations.
 * Borrows logic from [[java.lang.Integer.toHexString]] and [[java.util.HexFormat.toHexDigits]]
 */
def one() = {
  val sb = new java.lang.StringBuilder(32)
  sb.append("illegal character '\\u")
  val hexFormat = java.util.HexFormat.of()
  // computes bytes (min 2) from number of non-zero bits
  var numBytes = 2.max((Integer.SIZE - Integer.numberOfLeadingZeros(ch) + 7) / 8)
  while (numBytes > 0) {
    numBytes -= 1
    val curByte = (ch >>> (8 * numBytes)) & 0xf
    sb.append(hexFormat.toHighHexDigit(curByte))
    sb.append(hexFormat.toLowHexDigit(curByte))
  }
  sb.append('\'')
  curr.setInvalidToken(sb.toString)
}

/**
  * Prints 2 bytes if `ch` fits, or all 4 otherwise.
  * Has an extra memory allocation from `toHexDigits`.
  */
def two() = {
  val hexFormat = java.util.HexFormat.of()
  val output =
    if (Integer.numberOfLeadingZeros(ch) < 16)
      hexFormat.toHexDigits(ch)
    else
      hexFormat.toHexDigits(ch.toShort)
  curr.setInvalidToken("illegal character '\\u" + output + "'")
}
```